prow/plugins/trigger/trigger.go

```diff
@@ -153,3 +156,97 @@ func isUserTrusted(ghc githubClient, user, trustedOrg, org string) (bool, error)
 	}
 	return orgMember, nil
 }
+
+func runOrSkipRequested(c client, pr *github.PullRequest, requestedJobs []config.Presubmit, forceRunContexts map[string]bool, body, eventGUID string) error {
+	org := pr.Base.Repo.Owner.Login
+	repo := pr.Base.Repo.Name
+	number := pr.Number
+
+	baseSHA, err := c.GitHubClient.GetRef(org, repo, "heads/"+pr.Base.Ref)
+	if err != nil {
+		return err
+	}
+
+	getChanges := fileChangesGetter(c.GitHubClient, org, repo, number)
+	// shouldRun indicates if a job should actually run.
+	shouldRun := func(j config.Presubmit) (bool, error) {
+		if !j.RunsAgainstBranch(pr.Base.Ref) {
+			return false, nil
+		}
+		if j.RunIfChanged == "" || forceRunContexts[j.Context] || j.TriggerMatches(body) {
+			return true, nil
+		}
+		changes, err := getChanges()
```

Comment:
Why is this called per check? Will this race / give inconsistent results if the PR is updated while the plugin is trying to determine what to run on the last push?

Reply:
The getChanges closure has a cache, so this will fetch changes at most once.

Reply:
Ah, that diff was hard to read. Why structure it this way if the cache is per closure? Why not just list it before declaring this anonymous function?

Reply:
We don't want to always list the changes because we might not need to know what files were changed to properly trigger testing. We only have to fetch the changes if there is at least one RunIfChanged job that is not being force run (due to a `/retest` after a failure or because it is explicitly triggered with `/test foo`).

Reply:
OK. Maybe we can document that in a follow-up. The structure here is a bit complex and a "why" comment might help the next dev in this file :)